### ISSUE TITLE
[Bug]: Invocations cannot be uploaded when using dbt-snowflake version 1.5.0

### ISSUE THREAD

**Issue by mrsvllmr** [brooklyn-data/dbt_artifacts#296]
## Overview

We've updated dbt-snowflake to version 1.5.0 and now get the following error message:
`Object of type WarnErrorOptions is not JSON serializable`

It occurs in the course of uploading the invocations (package dbt_artifacts) on-run-end and ends with `Database error while running on-run-end`  and the log message above.

## How to reproduce

- update dbt-snowflake to version 1.5.0
- on-run-end: "{{ dbt_artifacts.upload_results(results) }}"
- run a dbt job

## Expected behaviour

As usual, upload the invocations after the job is finished.

## Environment

Results of running `dbt --version`:

```sh
 Core:
  - installed: 1.5.0
  - latest:    1.5.0 - Up to date!

Plugins:
  - snowflake: 1.5.0 - Up to date!
```

Please paste the contents of your `packages.yml` file here:

```yml
packages:
  - package: dbt-labs/dbt_utils
    version: 1.0.0

  - package: elementary-data/elementary
    version: 0.6.12

  - package: Snowflake-Labs/dbt_constraints
    version: 0.6.0

  - package: calogica/dbt_expectations
    version: [">=0.8.0", "<0.9.0"]

  - package: get-select/dbt_snowflake_monitoring
    version: 1.5.0

  - git: "https://github.com/svenrudolph/dbt_artifacts.git"
    revision: 2.2.2.10
```

Would this be solved when using 2.3.0 or is this an open issue (dbt-snowflake was just re-released yesterday)?

Logs:
`07:18:43.214 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43  Uploading invocations

07:18:43.240 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43  Database error while running on-run-end

07:18:43.241 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43  Connection 'master' was left open.

07:18:43.243 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43  On master: Close

07:18:43.338 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43  Connection 'list_stg_prd_dwh' was left open.

07:18:43.340 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43  On list_stg_prd_dwh: Close

07:18:43.447 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43  Connection 'model.douglas_datahub.dl_kafka__order_event_epp' was left open.

07:18:43.449 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43  On model.douglas_datahub.dl_kafka__order_event_epp: Close

07:18:43.540 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43  Connection 'list_stg_tst_dbt_snfl_fr' was left open.

07:18:43.541 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43  On list_stg_tst_dbt_snfl_fr: Close

07:18:43.652 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43  Connection 'list_dl_tst_dbt_pcm6' was left open.

07:18:43.653 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43  On list_dl_tst_dbt_pcm6: Close

07:18:43.731 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43

07:18:43.732 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43  Finished running 1 incremental model, 1 hook in 0 hours 0 minutes and 13.08 seconds (13.08s).

07:18:43.733 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43  Command end result

07:18:43.821 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43

07:18:43.822 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43  [31mCompleted with 1 error and 0 warnings:[0m
[0m07:18:43  

07:18:43.823 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43  [33mon-run-end failed, error:[0m
[0m07:18:43   Object of type WarnErrorOptions is not JSON serializable

07:18:43.824 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43
[0m07:18:43  Done. PASS=1 WARN=0 ERROR=1 SKIP=0 TOTAL=2

07:18:43.825 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43  Command `dbt build` failed at 07:18:43.824346 after 14.15 seconds

07:18:43.826 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43  Sending event: {'category': 'dbt', 'action': 'invocation', 'label': 'end', 'context': [<snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x7f063df09e70>, <snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x7f0636dc9cc0>, <snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x7f0634292c20>]}

07:18:43.827 | INFO    | Task run 'dbt build --select dl_kafka__order_event_epp-1' - [0m07:18:43  Flushing usage events

07:18:44.471 | ERROR   | Task run 'dbt build --select dl_kafka__order_event_epp-1' - Encountered exception during execution:
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 1545, in orchestrate_task_run
    result = await call.aresult()
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 181, in aresult
    return await asyncio.wrap_future(self.future)
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 218, in _run_async
    result = await coro
  File "/usr/local/lib/python3.10/site-packages/prefect_dbt/cli/commands.py", line 154, in trigger_dbt_cli_command
    result = await shell_run_command.fn(command=command, **shell_run_command_kwargs)
  File "/usr/local/lib/python3.10/site-packages/prefect_shell/commands.py", line 114, in shell_run_command
    raise RuntimeError(msg)
RuntimeError: Command failed with exit code 1:
[0m07:18:43  Flushing usage events

07:18:44.612 | ERROR   | Task run 'dbt build --select dl_kafka__order_event_epp-1' - Finished in state Failed('Task run encountered an exception: RuntimeError: Command failed with exit code 1:\n\x1b[0m07:18:43  Flushing usage events\n\n')
07:18:44.622 | ERROR   | Flow run 'master_order_event_subflow for batch_id (128372,)' - Encountered exception during execution:
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 673, in orchestrate_flow_run
    result = await flow_call.aresult()
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 181, in aresult
    return await asyncio.wrap_future(self.future)
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 194, in _run_sync
    result = self.fn(*self.args, **self.kwargs)
  File "/opt/prefect/flows/dbt/kafka_order/master_order_event.py", line 113, in master_order_event_worker
    master_order_event = dbt_snowflake.dbt_cli(f"{dbt_cmd_dl}")
  File "/usr/local/lib/python3.10/site-packages/prefect/events/instrument.py", line 73, in inner
    raise exc
  File "/usr/local/lib/python3.10/site-packages/prefect/events/instrument.py", line 70, in inner
    return function(self, *args, **kwargs)
  File "/opt/prefect/datahub/blocks.py", line 102, in dbt_cli
    result = trigger_dbt_cli_command.with_options(name=dbt_command)(
  File "/usr/local/lib/python3.10/site-packages/prefect/tasks.py", line 474, in __call__
    return enter_task_run_engine(
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 976, in enter_task_run_engine
    return from_sync.wait_for_call_in_loop_thread(begin_run)
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/api.py", line 137, in wait_for_call_in_loop_thread
    return call.result()
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 173, in result
    return self.future.result(timeout=timeout)
  File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 451, in result
    return self.__get_result()
  File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 218, in _run_async
    result = await coro
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 1131, in get_task_call_return_value
    return await future._result()
  File "/usr/local/lib/python3.10/site-packages/prefect/futures.py", line 240, in _result
    return await final_state.result(raise_on_failure=raise_on_failure, fetch=True)
  File "/usr/local/lib/python3.10/site-packages/prefect/states.py", line 91, in _get_state_result
    raise await get_state_exception(state)
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 1545, in orchestrate_task_run
    result = await call.aresult()
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 181, in aresult
    return await asyncio.wrap_future(self.future)
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 218, in _run_async
    result = await coro
  File "/usr/local/lib/python3.10/site-packages/prefect_dbt/cli/commands.py", line 154, in trigger_dbt_cli_command
    result = await shell_run_command.fn(command=command, **shell_run_command_kwargs)
  File "/usr/local/lib/python3.10/site-packages/prefect_shell/commands.py", line 114, in shell_run_command
    raise RuntimeError(msg)
RuntimeError: Command failed with exit code 1:
[0m07:18:43  Flushing usage events

07:18:44.773 | ERROR   | Flow run 'master_order_event_subflow for batch_id (128372,)' - Finished in state Failed('Flow run encountered an exception. RuntimeError: Command failed with exit code 1:\n\x1b[0m07:18:43  Flushing usage events\n\n')
07:18:44.812 | ERROR   | prefect.engine - Engine execution of flow run '12d81cd8-74f4-43f7-81db-23b2aeb19d6a' exited with unexpected exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 2131, in <module>
    enter_flow_run_engine_from_subprocess(flow_run_id)
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 202, in enter_flow_run_engine_from_subprocess
    ).result()
  File "/usr/local/lib/python3.10/site-packages/prefect/client/schemas.py", line 107, in result
    return get_state_result(self, raise_on_failure=raise_on_failure, fetch=fetch)
  File "/usr/local/lib/python3.10/site-packages/prefect/states.py", line 76, in get_state_result
    return _get_state_result(state, raise_on_failure=raise_on_failure)
  File "/usr/local/lib/python3.10/site-packages/prefect/utilities/asyncutils.py", line 260, in coroutine_wrapper
    return call()
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 245, in __call__
    return self.result()
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 173, in result
    return self.future.result(timeout=timeout)
  File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 451, in result
    return self.__get_result()
  File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 218, in _run_async
    result = await coro
  File "/usr/local/lib/python3.10/site-packages/prefect/states.py", line 91, in _get_state_result
    raise await get_state_exception(state)
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 673, in orchestrate_flow_run
    result = await flow_call.aresult()
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 181, in aresult
    return await asyncio.wrap_future(self.future)
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 194, in _run_sync
    result = self.fn(*self.args, **self.kwargs)
  File "/opt/prefect/flows/dbt/kafka_order/master_order_event.py", line 113, in master_order_event_worker
    master_order_event = dbt_snowflake.dbt_cli(f"{dbt_cmd_dl}")
  File "/usr/local/lib/python3.10/site-packages/prefect/events/instrument.py", line 73, in inner
    raise exc
  File "/usr/local/lib/python3.10/site-packages/prefect/events/instrument.py", line 70, in inner
    return function(self, *args, **kwargs)
  File "/opt/prefect/datahub/blocks.py", line 102, in dbt_cli
    result = trigger_dbt_cli_command.with_options(name=dbt_command)(
  File "/usr/local/lib/python3.10/site-packages/prefect/tasks.py", line 474, in __call__
    return enter_task_run_engine(
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 976, in enter_task_run_engine
    return from_sync.wait_for_call_in_loop_thread(begin_run)
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/api.py", line 137, in wait_for_call_in_loop_thread
    return call.result()
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 173, in result
    return self.future.result(timeout=timeout)
  File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 451, in result
    return self.__get_result()
  File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 218, in _run_async
    result = await coro
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 1131, in get_task_call_return_value
    return await future._result()
  File "/usr/local/lib/python3.10/site-packages/prefect/futures.py", line 240, in _result
    return await final_state.result(raise_on_failure=raise_on_failure, fetch=True)
  File "/usr/local/lib/python3.10/site-packages/prefect/states.py", line 91, in _get_state_result
    raise await get_state_exception(state)
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 1545, in orchestrate_task_run
    result = await call.aresult()
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 181, in aresult
    return await asyncio.wrap_future(self.future)
  File "/usr/local/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 218, in _run_async
    result = await coro
  File "/usr/local/lib/python3.10/site-packages/prefect_dbt/cli/commands.py", line 154, in trigger_dbt_cli_command
    result = await shell_run_command.fn(command=command, **shell_run_command_kwargs)
  File "/usr/local/lib/python3.10/site-packages/prefect_shell/commands.py", line 114, in shell_run_command
    raise RuntimeError(msg)
RuntimeError: Command failed with exit code 1:
[0m07:18:43  Flushing usage events`

**Comment by ajbosco** [comment]
This is happening to us as well.

**Comment by keolina-in** [comment]
This is also happening to us, was this supposed to be fixed in this?: https://status.getdbt.com/incidents/spwbbm1yc85r

**Comment by mikkosulonen** [comment]
> This is also happening to us, was this supposed to be fixed in this?: https://status.getdbt.com/incidents/spwbbm1yc85r

Nope. That fix was for an issue where the Cloud UI failed to start with an error: `module 'dbt.flags' has no attribute 'SEND_ANONYMOUS_USAGE_STATS'`

[dbt Slack thread about that issue](https://getdbt.slack.com/archives/CMZ2V0X8V/p1683003956404509)

**Comment by seub** [comment]
According to the dbt doc, the manifest version was updated to v9 (see [here](https://docs.getdbt.com/guides/migration/versions/upgrading-to-v1.5#for-consumers-of-dbt-artifacts-metadata)). I assume this is why it breaks this package's current version.

**Comment by TomSchiro** [comment]
Origina error:
22:35:40  on-run-end failed, error:
22:35:40   Object of type WarnErrorOptions is not JSON serializable

Seems be coming from line 66 in the macro upload_invocations.sql and this is the value
'{{ tojson(invocation_args_dict) | replace('\\', '\\\\') }}'

I simply replace line 66 with was in line 67 which is '{{ tojson(dbt_metadata_envs) }}'

The json for what I believe is causing the issue is from warn_error_options and the structure of it is 2 empty arrays inside a dictionary.  

I also output this to the log file since I was curious of the actual value
'warn_error_options': WarnErrorOptions(include=[], exclude=[])

If not known already hope this helps.  You guys did great with this.

**Comment by c42-josh-carlson** [comment]
For those watching, I was able to get this to run by over-riding the invocations macro in my project and replace line 66 in the upload_invocations macro...

`'{{ tojson(invocation_args_dict) | replace('\\', '\\\\') }}', {# invocation_args #}`

With the following... 

```
{% set keys_to_find = ['event_buffer_size', 'indirect_selection', 'no_print', 'partial_parse', 'printer_width', 'profiles_dir', 'quiet', 'rpc_method', 'select', 'send_anonymous_usage_stats', 'static_parser', 'use_colors', 'version_check', 'which', 'profile', 'defer', 'exclude', 'full_refresh', 'write_json', 'resource_types', 'state', 'target', 'cache_selected_only', 'compile'] %}
{% set new_invoke_args = {} %}
{% for key in keys_to_find %}
    {% if key in invocation_args_dict | list  %}
        {% do new_invoke_args.update({key: invocation_args_dict[key]}) %}
    {% endif %}
{% endfor %}
```

Essentially, there are a number of keys that have been added to invocation_args_dict that do not return as a string or a valid JSON. While 'warn_error_options': WarnErrorOptions(include=[], exclude=[]) is one of them, even when you adjust that there are two other keys `'select'` and `exclude` which also cause parsing errors. I found the best solution was to force the invocation args to only upload the keys that were there before. An admittedly hacky solution but one that is fine enough for now to get dbt_artifacts working with DBT 1.5.

**Comment by glsdown** [comment]
Hi all, thanks for contributing (and thanks @c42-josh-carlson for your workaround here). We are going to prioritise this fix to get everything working with v1.5.0. I will keep you posted.

**Comment by ajbosco** [comment]
@glsdown anything we can do to help with the release? We're anxious to get this working with dbt-core 1.5.0 too!

**Comment by glsdown** [comment]
Hello everyone here. Apologies for how long it has taken to get this release out. We had some issues with the dbt-databricks connector which was causing a lot of headaches. We have been able to work around it now though whilst they resolve the underlying issue, so you should see the fix for this in v2.4.0. Please do let me know if it is not working for whatever reason though.
